winsup/cygwin/include/cygwin/exit_process.h

```diff
@@ -0,0 +1,281 @@
+
+  si.cb = sizeof (si);
+  si.dwFlags = STARTF_USESHOWWINDOW | STARTF_USESTDHANDLES;
+  si.wShowWindow = SW_HIDE;
+  si.lpTitle = title;
+  si.hStdInput = si.hStdError = si.hStdOutput = INVALID_HANDLE_VALUE;
+
+  /* Create a new hidden process. */
+  BOOL x = CreateProcessW (NULL, cmd, NULL, NULL, TRUE,
+			   CREATE_NO_WINDOW | CREATE_NEW_PROCESS_GROUP,
+			   NULL, NULL, &si, &pi);
+  CloseHandle(h[1]);
+  if (!x)
+    {
+      CloseHandle(h[0]);
+      return NULL;
+    } else {
+    /* Wait for the process to complete for 10 seconds */
+      WaitForSingleObject(pi.hProcess, 10000 );
+    }
+
+  if (!GetExitCodeProcess(pi.hProcess, process_exit))
+    return NULL;
+
+  CloseHandle(pi.hThread);
+  CloseHandle(pi.hProcess);
+  CloseHandle(h[0]);
+
+  return process_exit;
+}
+
+static int current_is_wow = -1;
+static int is_32_bit_os = -1;
+
+static BOOL get_wow(HANDLE process, BOOL& is_wow, int& bitness)
+{
+  if (is_32_bit_os == -1)
+    {
+      SYSTEM_INFO info;
+
+      GetNativeSystemInfo(&info);
+      if (info.wProcessorArchitecture == 0)
+	is_32_bit_os = 1;
+      else if (info.wProcessorArchitecture == 9)
+	is_32_bit_os = 0;
+      else
+	is_32_bit_os = -2;
+    }
+
+  if (current_is_wow == -1 &&
+      !IsWow64Process (GetCurrentProcess (), &current_is_wow))
+    current_is_wow = -2;
+
+  if (is_32_bit_os == -2 || current_is_wow == -2)
+    return FALSE;
+
+  if (!IsWow64Process (process, &is_wow))
+    return FALSE;
+
+  bitness = is_32_bit_os || is_wow ? 32 : 64;
+  return TRUE;
+}
+
+/**
+ * Terminates the process corresponding to the process ID
+ *
+ * This way of terminating the processes is not gentle: the process gets
+ * no chance of cleaning up after itself (closing file handles, removing
+ * .lock files, terminating spawned processes (if any), etc).
+ */
+static int
+exit_one_process(HANDLE process, int exit_code, DWORD pid)
+{
+  LPTHREAD_START_ROUTINE address = NULL;
+  int signo = exit_code & 0x7f;
+  BOOL is_wow;
+  int bitness;
+  switch (signo)
+  {
+  case SIGINT:
+  case SIGQUIT:
+    if (get_wow(process, is_wow, bitness) &&
+        0 /* exit code 0 = success */ == kill_via_console_helper(bitness, L"CtrlRoutine",
+                                                                 signo == SIGINT ? CTRL_C_EVENT : CTRL_BREAK_EVENT, pid))
+      return 0;
+    /* fall-through */
+  case SIGTERM:
+    if (get_wow(process, is_wow, bitness) &&
+        0 /* exit code 0 = success */ == kill_via_console_helper(bitness, L"ExitProcess", exit_code, pid))
+      return 0;
+    break;
+  default:
+    break;
+  }
+
+  return int(TerminateProcess (process, exit_code));
+}
+
+#include <tlhelp32.h>
+#include <unistd.h>
+
+/**
+ * Terminates the process corresponding to the process ID and all of its
+ * directly and indirectly spawned subprocesses using the
+ * exit_one_process() function.
+ */
+static int
+exit_process_tree(HANDLE main_process, int exit_code)
```

Comment:
The idea of `exit_process_tree()` is this: contrary to POSIX' `kill()` (at least with certain signals), `TerminateProcess()` does not leave the killed processes _any_ way to call `atexit()`-registered handlers. But those would be the ones that kill child processes. Hence we need to enumerate the process tree and terminate the children, too.

However, sending a CtrlEvent is different: it _does_ allow `atexit()`-registered handlers to run.

Therefore, I would contend that we should not even call `exit_process_tree()` by default, but something like `exit_process()` (which either sends a CtrlEvent to the specified process or injects `ExitProcess()`, and that process _only_, or terminates the process tree only if `TerminateProcess()` is used).

Reply:
And yes, I know, this is my fault because I did not change the name after realizing that the `atexit()`-registered handlers weren't run upon `TerminateProcess()` and looking for a better solution (and finding one, first with injecting `ExitProcess()`, and then even injecting `CtrlRoutine()`).

Reply:
I guess the solution is actually to turn things around, letting `exit_process_tree()` call `TerminateProcess()` instead of `exit_one_process()` (and losing the `int exit_code` parameter). Then, renaming `exit_one_process()` to `exit_process()`, and calling `exit_process_tree()` instead of `TerminateProcess()` when falling back.

Reply:
> and losing the `int exit_code` parameter

what should be passed to `TerminateProcess` then?

Reply:
Does https://github.com/msys2/msys2-runtime/pull/31/commits/efc36e4202b8a62acb1183e888b5aafb70f6bade look good?